# Grundlagen von Jupyter Notebooks
#### Marcel Lüthi, Andreas Morel-Forster<br/>Departement Mathematik und Informatik, Universität Basel

In diesem Notebook schauen wir uns vier grundlegende Datentypen von Java an. Sie lernen:
- den Wertebereich der Datentypen.
- wie Sie die Datentypen zwischeneinander konvertieren können.
- was Sie beachten müssen wenn es um die Genauigkeit geht.

### Datentypen für ganze Zahlen und Gleitkommazahlen

Sie haben den Unterschied zwischen ganzen Zahlen und Gleitkommazahlen in Java kennen gelernt. Dafür haben Sie bisher `int` und `double` verwendet. In Java gibt es aber noch den Datentpen `long` für ganze Zahlen, dessen Wertebereich grösser ist als der von `int`. Bei den Gleitkommazahlen gibt es noch den Datentypen `float`, welche die Zahlen weniger genau darstellt als `double`.

Für unsere Vorlesung gilt:

- Ganze Zahlen `int`, `long` wenn der Wertebereich von `int` nicht gross genug ist.
- Gleitkommazahlen `double`, `float` wenn der Speicher zu begrenzt ist.
- (Nicht in der Vorlesung `byte` und `short`, welche ganze Zahlen nur in einem kleineren Bereich als `int` darstellen können.)

#### Ausdrücke

Um einen Zahlwert als `long` zu kennzeichnen, folgt nach der ganzen Zahl der Buchstabe `l` (kleines `L`).
```java
1l
987654321l
```
Um einen Zahlenwert als `float` zu kennzeichnen, folgt nach dem Zahlwert der Buchstabe `f`.
```java
1f
-2.345f
3.1415e-3f
```

#### Speicherbedarf

Der Speicherbedarf wird gemessen, indem man die zum Speichern benötigten Bits, als die Anzahl Nullen und Einsen, angibt. Den Speicherbedarf können Sie mit Hilfe von in Java definierten Konstanten heraus finden.

Die Konstanten welche den Speicherbedarf angeben sind:

| Typ | Konstante |
| --- | --- |
| int | Integer.SIZE |
| long | Long.SIZE |
| float | Float.SIZE |
| double | Double.SIZE |


In [81]:
public class MemoryFootprint {
    public static void main(String[] args) {
        System.out.println("int belegt " + Integer.SIZE + " Bits im Speicher.");
        System.out.println("long belegt " + Long.SIZE + " Bits im Speicher.");
        System.out.println("float belegt " + Float.SIZE + " Bits im Speicher.");
        System.out.println("double belegt " + Double.SIZE + " Bits im Speicher.");
    }
}

In [82]:
MemoryFootprint.main(new String[0]);

int belegt 32 Bits im Speicher.
long belegt 64 Bits im Speicher.
float belegt 32 Bits im Speicher.
double belegt 64 Bits im Speicher.


#### Wertebereich

Genau wie für den Speicherverbrauch gibt es in Java definierte Konstanten welche den Wertebereich der jeweiligen Datentypen angeben. Vorsicht: Die Konstanten mit dem selben Namen können für unterschiedliche Typen etwas anderes bedeuten.

Konstanten welche den Wertebereich angeben sind:

| Typ | Konstante | Bedeutung |
| --- | --- | --- |
| int | Integer.MAX_VALUE | grösster, darstellbarer Wert |
| int | Integer.MIN_VALUE | kleinster, darstellbarer Wert |
| long | Long.MAX_VALUE | grösster, darstellbarer Wert |
| long | Integer.MIN_VALUE | kleinster, darstellbarer Wert |
| float | Float.MAX_VALUE | grösster, darstellbarer Wert |
| float | Float.MIN_VALUE | kleinster, positiver, darstellbarer Wert |
| double | Double.MAX_VALUE | grösster, darstellbarer Wert |
| double | Double.MIN_VALUE | kleinster, positiver, darstellbarer Wert |

<div class="alert alert-block alert-info">
    Das Minimum der darstellbaren Gleitkommazahlen finden Sie als das negierte Maximum.
</div>

In [87]:
public class Range {
    public static void main(String[] args) {
        System.out.println("int stellt Zahlen von " +
            Integer.MIN_VALUE + " bis " +
            Integer.MAX_VALUE + " dar. Dies ist in etwa +- 2 Milliarden.");
        System.out.println("long stellt Zahlen von " +
            Long.MIN_VALUE + " bis " +
            Long.MAX_VALUE + " dar. Dies ist in etwa +- 9 Trillionen.");
        System.out.println("float stellt Zahlen von " +
            (-Float.MAX_VALUE) + " bis " +
            Float.MAX_VALUE + " dar.");
        System.out.println("double stellt Zahlen von " +
            (-Double.MAX_VALUE) + " bis " +
            Double.MAX_VALUE + " dar.");
        
        System.out.println(Float.MIN_VALUE + " ist nur etwas grösser als 0.0f.");
        System.out.println(Double.MIN_VALUE + " ist nur etwas grösser als 0.0.");
    }
}

In [88]:
Range.main(new String[0]);

int stellt Zahlen von -2147483648 bis 2147483647 dar. Dies ist in etwa +- 2 Milliarden.
long stellt Zahlen von -9223372036854775808 bis 9223372036854775807 dar. Dies ist in etwa +- 9 Trillionen.
float stellt Zahlen von -3.4028235E38 bis 3.4028235E38 dar.
double stellt Zahlen von -1.7976931348623157E308 bis 1.7976931348623157E308 dar.
1.4E-45 ist nur etwas grösser als 0.0f.
4.9E-324 ist nur etwas grösser als 0.0.


#### Welches ist der richtige Typ?

Welchen Typen würden Sie bei den folgenden Aufgabenstellungen:
- "Speichern Sie den Geldbetrag eines Bankkunden."
- "Speichern Sie das Gewicht einer Person."
- "Zählen Sie die Häuser in der Stadt Basel."
- "Speichern Sie die Zahl Pi."
- "Zählen Sie alle Menschen auf dieser Erde."

Für diesen Kurs gilt in der Regel: Wenn nichts spezielles steht, nehmen Sie entweder `double` oder `int`. Dabei verwenden wir `double` für Messgrössen und `int` für Zählgrössen.

#### Miniaufgabe

Berechnen Sie die Lösung für die folgende Aufgabe und geben Sie die Zahl aus.
- Sie wollen einen Gartenzaun der Länge 20m bauen.
- Ein Element hat die Länge 1.48m.
- Wieviele Elemente müssen Sie kaufen?

In [6]:
int meters = 20;
double stone = 0.3;
double d = meters/stone;
long l = (long) Math.ceil(meters/stone);
System.out.println(d + " " + l);

66.66666666666667 67


#### Umwandeln zwischen Datentypen

Auf den Wertebereiche der Datentypen bezogen gilt:
$$ int \supset long \supset float \supset double$$

Bei der Zuweisung können Sie Werte jeweils an den gleichen oder an einen *mächtigeren* Datentypen zuweisen. In die andere Richtung benötigen Sie *Type-Casting*.

#### Type-Casting

Beim Type-Casting schreibt man den Typ der Variablen in runden Klammern vor den Ausdruck der umgewandelt werden soll.

Achtung: Das Umwandeln mit Type-Casting kommt meist auch mit einer gewissen Ungenauigkeit.

Schema
```
(<TYPE>) <AUSDRUCK>
```

Beispiele
```java
(int) 2.0
(float) 3.1415
(long) 5.543
```

#### Miniübung

- Überlegen Sie sich für jede der Variablen Initialiserungen was nicht stimmt.
- Versuchen Sie die Zelle auszuführen.
- Lesen Sie sich die Fehlermeldungen des Kompilers durch.
- Ändern Sie die Zeilen mit Hilfe von Type-Casting so, dass sie den Code ausführen können.
- Geben Sie am Ende die Zahlen `l` und `r` aus und diskutieren Sie das Ergebnis.

In [80]:
public class TypeCasts {
    public static void main(String[] args) {
        int i = 2.0;
        float f = 4 * Math.PI;
        long l = f;
        int r = Math.round(4 * Math.PI);
        System.out.println(l + " " + r);
    }
}

CompilationException: 

In [79]:
TypeCasts.main(new String[0]);

12 13


#### Genauigkeit von Zahlen

Die Genauigkeit von Zahlen im Computer ist Beschränkt. Nicht nur beim Type-Casting können also Fehler entstehen.

#### Miniübung

Berechnen Sie die Summe von 

In [97]:
public class Summation {
    public static void main(String[] args) {
        double result = 0.1 + 0.1 + 0.1;
        System.out.println("Das Ergebnis ist nicht 0.3 sondern " + result + ".");
    }
}

In [93]:
Summation.main(new String[0]);

Das Ergebnis ist nicht 0.3 sondern 0.30000000000000004


Mathematisch gibt die folgende Formel für positive, ganze Zahlen immer 1:
$$\sum\limits_{i=1}^{N}\frac{1}{N}$$

In [94]:
public class Summation2{
    public static void main(String[] args) {
        double result = 1.0 / 6.0 + 1.0 / 6.0 + 1.0 / 6.0 + 1.0 / 6.0 + 1.0 / 6.0 + 1.0 / 6.0;
        System.out.println("Das Ergebnis ist nicht 1 sonder " + result + ".");
    }
}

In [96]:
Summation2.main(new String[0]);

Das Ergebnis ist nicht 1 sonder 0.9999999999999999.


Dies zeigt, der Computer rechnet dies aus und lösst die Aufgabe nicht irgendwie algebraisch.

Nachfolgend sehen Sie ein Beispiel einer Aufgabe in einem Lernsetting. Es geht darum drei Zahlen zu Summieren.

In [98]:
public class Summation3 {
    public static void main(String[] args) {
        double a = 13.0;
        double b = Math.PI;
        double c = 21.45;

        double solution_student = a + b + c;
        System.out.println("The solution of the student is: " + solution_student);

        double solution_teacher = a + c + b;
        System.out.println("The solution of the teacher is: " + solution_teacher);
    }
}

In [99]:
Summation3.main(new String[0]);

The solution of the student is: 37.59159265358979
The solution of the teacher is: 37.5915926535898


Diese Beobachtungen zeigen, dass man bei Gleitkommazahlen immer aufpassen muss mit der Genauigkeit. Dies wird später im Kurs vorallem auch beim Testen von Programmen erneut wichtig.

#### Beobachtungen

- Zahlen vom Typ `float` werden mit 32 Bits dargestellt. Mit 32 Bits lassen sich $2^{32}$ unterschiedliche Werte darstellen.
- Der Wertebereich ist ungefähr $-3.4*10^{38}$ bis $-3.4*10^{38}$, das ist ein Intervall der Grösse $2^{162}$.
- Der absolute Abstand zweier benachbarter, darstellbaren Zahlen rund um 0.0f ist kleiner als für Zahlen nahe der Grenzen des darstellbaren Wertebereichs.
- Intuition: In der wissenschaftlichen Schreibweise von Zahlen (z.B.: $3.14159*10^{7}$), hat `float` eine Genauigkeit von 6 oder 7 signifikante Dezimalstellen, für `double` sind es 14 bis 15 signifikante Dezimalstellen.

Interessierte finden eine ausführlichere und detailliertere Diskussion rund um die Darstellung und Genauigkeit von `float` [hier](https://introcs.cs.princeton.edu/java/91float/).